### PR TITLE
Analyzer planner fixes before enable by default

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+                output_nodes_mapped_to_input.emplace(output_node->result_name);
+
+                auto output_names_it = input_node_to_output_names.find(node_without_alias->result_name);
+                if (output_names_it == input_node_to_output_names.end())
+                {
+                    auto [insert_it, _] = input_node_to_output_names.emplace(node_without_alias->result_name, std::vector<std::string>());
+                    output_names_it = insert_it;
+                }
+
+                output_names_it->second.push_back(output_node->result_name);
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp:130]: Remove copy-paste

File: src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+        storage_metadata,
+        queried_columns,
+        storage.supportedPrewhereColumns(),
+        &Poco::Logger::get("QueryPlanOptimizePrewhere")};
+
+    auto optimize_result = where_optimizer.optimize(filter_step->getExpression(),
+        filter_step->getFilterColumnName(),
+        read_from_merge_tree->getContext(),
+        is_final);
+    if (!optimize_result)
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp:176]: `!= std::noopt`

File: src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+      * 2. Prewhere filter expressions.
+      *
+      * There can be cases when all expressions are moved to PREWHERE, but it is not
+      * enough to produce required filter output columns.
+      *
+      * Example: SELECT (a AND b) AS cond FROM test_table WHERE cond AND c;
+      * In this example conditions a, b, c can move to PREWHERE, but PREWHERE will not contain expression and(a, b).
+      * It will contain only a, b, c, and(a, b, c) expressions.
+      *
+      * In such scenario we need to create additional filter expressions after PREWHERE.
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp:207]: Add more info

File: src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+
+void matchDAGOutputNodesWithHeader(ActionsDAGPtr & actions_dag, const Block & expected_header)
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp:21]: matchDAGOutputNodesOrderWithHeader

File: src/Planner/Planner.cpp [hunk 5]
```diff
@@ -79,26 +79,12 @@ namespace ErrorCodes
...
 
 /** ClickHouse query planner.
   *
-  * TODO: Support JOIN with JOIN engine.
-  * TODO: Support VIEWs.
-  * TODO: JOIN drop unnecessary columns after ON, USING section
-  * TODO: Support RBAC. Support RBAC for ALIAS columns
-  * TODO: Support PREWHERE
-  * TODO: Support DISTINCT
-  * TODO: Support trivial count optimization
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Planner/Planner.cpp:93]: Updated.

File: src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp]: Updated.

File: src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp (hunk 3 above)
Comment (kitaisreal) [reply src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp]: Updated.

File: src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp (hunk 2 above)
Comment (kitaisreal) [reply src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp]: Updated.

**Comment by alexey-milovidov** [comment]
There was one error; please resubmit.

**Pull request merged**